In [1]:
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq
import matplotlib.pyplot as plt
import gc
import pandas as pd
import numpy as np
from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_val_score, RandomizedSearchCV, cross_val_predict, cross_validate
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report,precision_score, RocCurveDisplay, f1_score, recall_score, roc_curve, auc, ConfusionMatrixDisplay, make_scorer, balanced_accuracy_score 
from sklearn.ensemble import HistGradientBoostingClassifier, RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder, StandardScaler, RobustScaler
from sklearn.impute import SimpleImputer
import timeit
import time
import xgboost as xgb
import lightgbm as lgb
from sklearn.model_selection import TimeSeriesSplit, StratifiedGroupKFold
import mlflow
from sklearn.utils.class_weight import compute_sample_weight
import pathlib 

In [2]:
# Set up mlflow connection with database
mlflow.set_tracking_uri("postgresql://postgres:mars2006@localhost:5432/mlflow_oil_logging")

mlflow.set_experiment("Oil_Well_Fault_Detection")

<Experiment: artifact_location='/Users/Tari/Desktop/Tari/oil_dataset/predictive_maintenance_dataset/notebooks/mlruns/1', creation_time=1789848169371, effective_trace_archival_retention=None, experiment_id='1', last_update_time=1789848169371, lifecycle_stage='active', name='Oil_Well_Fault_Detection', tags={}, trace_location=None, workspace='default'>

In [3]:
mlflow.autolog(silent=True,disable=True)

In [5]:
pip install psycopg2

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for psycopg2: filename=psycopg2-2.9.13-cp314-cp314-macosx_12_0_arm64.whl size=133947 sha256=29f103889cc26dc60ea0a0e184e6ee51e4bae79e5ce7eeb6b3477f3b55b0df83
  Stored in directory: /Users/Tari/Library/Caches/pip/wheels/6d/13/0d/e7151d6c5695aab4ae1e34a9151233d8522ea9bdfda8ffddae
Successfully built psycopg2
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install --upgrade "mlflow>=3.1"

  Using cached alembic-1.20.0-py3-none-any.whl.metadata (7.3 kB)
  Using cached mako-1.4.1-py3-none-any.whl.metadata (2.9 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.7/11.7 MB 517.6 kB/s  0:00:22 eta 0:00:010:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 232.3 kB/s  0:00:15241.8 kB/s eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 145.8 kB/s  0:00:13131.9 kB/s eta 0:00:02
Using cached alembic-1.20.0-py3-none-any.whl (268 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.0/4.0 MB 642.2 kB/s  0:00:06a 0:00:01m eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 327.7 kB/s  0:00:03352.4 kB/s eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 277.8 kB/s  0:00:07288.2 kB/s eta 0:00:02
Using cached mako-1.4.1-py3-none-any.whl (80 kB)
  Attempting uninstall: wcwidth2;249;38;114m╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  2/45 [werkzeug]
    Found existing installation: wcwidth 0.2.1438;5;237m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [5]:
pip install pyarrow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 35.9/35.9 MB 5.4 MB/s  0:00:07 eta 0:00:010:00:01
Note: you may need to restart the kernel to use updated packages.


In [4]:
preprocessed_df = pd.read_csv('../cleaned_csv/cleaned_data_with_no_pdg.csv')

In [5]:
preprocessed_df["class"].value_counts()

class
5.0      352058
0.0      242892
101.0    175457
3.0      148625
1.0       96982
107.0     76535
108.0     43177
4.0       41585
105.0     35300
8.0       20163
2.0       11609
7.0        3625
102.0      2189
6.0         417
106.0       152
Name: count, dtype: int64

In [84]:
custom_weight = {
    0: 1,
    1: 1,
    2: 1,
    3: 1,
    4: 1,
    5: 1,
    6: 2,
    7: 1,
    8: 1,
    101: 1,
    102: 4,
    105: 1,
    106: 2,
    107: 1,
    108: 1
}

In [6]:
custom_weight_transient = {
    0:   1.0,
    101: 1.0,
    102: 2.0,
    105: 1.0,
    106: 4.0,
    107: 1.0,
    108: 1.0,
}

## Phase two : Define the actual ML Problem
### Train models using 3 datasets: 
* Current Dataset
* Dataset without class 3 and 4
* Dataset without permanent faults


In [5]:
preprocessed_df.isna().sum()

timestamp             0
P-TPT                 0
T-TPT                 0
P-MON-CKP             0
T-JUS-CKP             0
                     ..
P_MON_CKP_slope_60    0
P_TPT_std_60          0
T_TPT_std_60          0
all_stable            0
stability_streak      0
Length: 63, dtype: int64

In [42]:
#---------------------------------
# VISUALIZATION PURPOSES ONLY
#----------------------------------

# Split data into X and Y.
X = preprocessed_df.drop(columns=["class","Instance_id"])
X = X.set_index("timestamp")
y = preprocessed_df["class"]
group = preprocessed_df["Instance_id"]

# Use StratifiedGroupKFold to split data into training and testing.
sgkf = StratifiedGroupKFold(n_splits=2) # can be changed.

for i, (train_idx, test_idx) in enumerate(sgkf.split(X,y,preprocessed_df["Instance_id"])):
    print(f"\n--- Fold {i + 1} ---")
    X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
    y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]
    groups_train, groups_test = group.iloc[train_idx], group.iloc[test_idx]

    print(f"Train Group IDs: {np.unique(groups_train)} | Test Group IDs: {np.unique(groups_test)}")
    cleared_obj = gc.collect()
    
print(f"Cleared {cleared_obj} trapped objects from RAM")


--- Fold 1 ---
Train Group IDs: ['0_WELL-00001_20170201020207' '0_WELL-00001_20170201120124'
 '0_WELL-00001_20170218060218' '0_WELL-00001_20170218160135'
 '0_WELL-00001_20170218210311' '0_WELL-00001_20170219070031'
 '0_WELL-00001_20170219120021' '0_WELL-00001_20170219220332'
 '0_WELL-00001_20170422040000' '0_WELL-00001_20170424180042'
 '0_WELL-00001_20170425040207' '0_WELL-00001_20170521050000'
 '0_WELL-00001_20170521200000' '0_WELL-00001_20170522010000'
 '0_WELL-00001_20170522110000' '0_WELL-00001_20170522160000'
 '0_WELL-00001_20170523020012' '0_WELL-00001_20170523070000'
 '0_WELL-00001_20170523170000' '0_WELL-00001_20170524030000'
 '0_WELL-00001_20170524080000' '0_WELL-00001_20170524130000'
 '0_WELL-00001_20170525090000' '0_WELL-00001_20170525140000'
 '0_WELL-00001_20170525190000' '0_WELL-00001_20170526050000'
 '0_WELL-00001_20170526100000' '0_WELL-00001_20170527010000'
 '0_WELL-00001_20170527060000' '0_WELL-00001_20170527160000'
 '0_WELL-00001_20170527210000' '0_WELL-00001_2017052

In [7]:
def ConfusionMatrix(y_test,y_preds, model_name, desc, color="Blues"):
    # Plot confusion matrix
    labels = np.sort(np.unique(y_test))

    cfm = confusion_matrix(y_test, 
                           y_preds, 
                           labels=labels)
    
    disp = ConfusionMatrixDisplay(confusion_matrix=cfm, 
                                  display_labels=labels
                                  )
    
    fig, ax = plt.subplots(figsize=(14, 12))
    disp.plot(ax=ax, colorbar=True, cmap=color)

    plt.tick_params(axis="x", rotation=45)
    plt.title(f'{model_name} Confusion Matrix On Test Data')
    plt.tight_layout()

    filename = f"confusion_matrix_{model_name}_{desc}.png"

    plt.savefig(filename, dpi=130, bbox_inches='tight')

    plt.close(fig)

    return filename

In [8]:
def feature_importance(model,X, model_name, desc, i, feature_dir):

    sorted_indices = np.argsort(model.feature_importances_)[::-1][:25]
    
    # Reorder both the importances and the column names using those indices
    # (Assumes 'X' is your original pandas training DataFrame)
    sorted_importances = model.feature_importances_[sorted_indices]
    sorted_columns = X.columns[sorted_indices]
    
    # 3. Plot a horizontal bar chart
    fig, ax = plt.subplots(figsize=(14, 10))
    
    # Use ax.barh to map values to labels
    ax.barh(sorted_columns, sorted_importances, label="Feature Importances")
    
    ax.set_xlabel("Importance Value") 
    ax.set_ylabel("Features")  
    ax.set_title(f"{model_name} Feature Importance - Fold {i + 1}")       
    ax.invert_yaxis()                  # Puts the highest importance at the top
    ax.legend()
    plt.tight_layout()

    filename = (f"Feature_importance_{model_name}_{desc}_v{i + 1}.png")

    filepath = feature_dir / filename
    
    fig.savefig(
        filepath,
        dpi=150,
        bbox_inches="tight"
    )

    plt.close(fig)

    return filepath

In [39]:
#------------------------------------------#
# REUSABLE FUNCTION FOR MODEL TRAINING
#------------------------------------------#

def run_model(model,model_name,splits,X,y,group,scoring, sample_weight= None):
    start_time = time.perf_counter()
    print(f"Starting running time...")

    #--------- Create Pipeline ------#
    pipe = Pipeline([("simpleImputer", SimpleImputer(strategy="median")),
                      ("RobustScaler", RobustScaler()),
                      (f"{model_name}", model)])
    
    #--------- Instantiate split class -----#
    sgkf = StratifiedGroupKFold(n_splits=splits, shuffle=True, random_state=42)

    cv_splits = list(
    sgkf.split(X, y, group)
    )

       #--------- Pass sample weights to model.fit() -----#
    fit_params = None
    if sample_weight is not None:
        fit_params = {
            f"{model_name}__sample_weight":
            sample_weight
        }
    
    #--------- Execute Cross-Validation -------#
    print(f"Starting {model} cross-validation...")
    cv_scores = cross_validate(
        estimator=pipe,             
        X=X,                 
        y=y,                 
        # groups=group,       
        cv=cv_splits,            
        scoring=scoring,
        return_estimator=True,
        return_train_score=True,
        verbose=0,
        n_jobs=1
    )
 
    #--------- Create a Prediction Array for Comparing and for Confusion Matrix Evaluation -----#
    y_preds = cross_val_predict(estimator=pipe,
                                X=X,
                                y=y,
                                groups=group,
                                cv=sgkf,
                                n_jobs=1,
                                )
    
    #----------- Log results -----------#

    # ---- F1 MACRO ---- #
    print(f"Macro F1 Scores (Test): {cv_scores["test_f1_macro"]}")

    # ---- ACCURACY SCORE ---- #
    print(f"===============\n"
          f"Accuracy score of Test set: {cv_scores["test_accuracy"]}\n" 
          f"Accuracy score of Training set: {cv_scores["train_accuracy"]}\n"
          f"===============")

    # ---- BALANCED ACCURACY ---- #
    print(f"Balanced Accuracy score (Test): {cv_scores["test_balanced_accuracy"]}")

    # ---- RECALL MACRO ---- #
    print(f"Recall Macro score (Test): {cv_scores["test_recall_macro"]}")

    # ---- AVERAGE TEST SCORE USING F1 MACRO ---- #
    print(f"Average Test Score (Test): {(cv_scores["test_f1_macro"].mean() * 100):.2f}%")

    end_time = time.perf_counter()
    execution_time = end_time - start_time 
    print(f"Execution time: {execution_time} seconds")

    gc.collect()
    return (cv_scores, cv_scores["estimator"], y_preds, y)



## Model 1: LightGBM

### Experiment one: Dataset without class 3 and 4

In [11]:
# ========= CREATE A COPY OF THE ORIGINAL DATASET TO WORK WITH =========
df_copy_without_class_3_4 = preprocessed_df.copy()
df_copy_without_class_3_4 = df_copy_without_class_3_4.set_index("timestamp")

In [12]:
#------ Create Dataset without class 3 and 4------#
df_copy_without_class_3_4 = df_copy_without_class_3_4[~df_copy_without_class_3_4["class"].isin([3,4])]

In [13]:
# ======= INSTANTIATE FEATURE AND TARGET LABELS =======#
X_without_3_4 = df_copy_without_class_3_4.drop(columns=["class","Instance_id"])

y_without_3_4 = df_copy_without_class_3_4["class"]

group_without_3_4 = df_copy_without_class_3_4["Instance_id"]

In [14]:
params = {
    "n_estimators": 600,
    "learning_rate": 0.05,
    "num_leaves": 300,
    "max_depth": 12,
    "min_child_samples": 200,
    "class_weight": "balanced",
    "n_jobs": -1,
    "random_state": 42,
}


model = lgb.LGBMClassifier(**params)
model_name = "LightGBM"


splits = 3

scoring = {
    "f1_macro": "f1_macro",
    "accuracy": "accuracy",
    "recall_macro": "recall_macro",
    "balanced_accuracy": "balanced_accuracy"
}

description = "without_class_3_and_4_dataset"

# -------- CREATE FOLDER TO STORE FEATURE IMPORTANCES ------ #
feature_dir = pathlib.Path(f"Feature_importance_{model_name}")
feature_dir.mkdir(parents=True, exist_ok=True)

with mlflow.start_run(run_name="Lightgbm_tuned_v1_without_class_3_and_4_dataset", description="Trained without class 3 and 4"):
    cv_scores, estimator, y_preds, y = run_model(model,
                                     model_name,
                                     splits,
                                     X_without_3_4,
                                     y_without_3_4,
                                     group_without_3_4,
                                     scoring
                                     )

    #------- Create Feature Importance for each fold -------#
    for i in range(splits):
        model = estimator[i].named_steps[f"{model_name}"]
        feature_path = feature_importance(model,X_without_3_4,model_name, description,i, feature_dir)

        importance_df = pd.DataFrame({
            "Feature": X_without_3_4.columns,
            "Importance": model.feature_importances_
        }).sort_values("Importance", ascending=False)

        csv_path = feature_dir / f"lightgbm_feature_importances_v{i+1}.csv"

        # Save ALL 72 features
        importance_df.to_csv(
            csv_path,
            index=False
        )

        #------- Save Feature Importance plot in Mlflow run ------#
        mlflow.log_artifact(str(feature_path), artifact_path="feature_importance")
        
    # ---------  Confusion Matrix plot ------ #
    cm_path = ConfusionMatrix(y, y_preds, model_name, description)
    mlflow.log_artifact(cm_path)
        
    mlflow.log_params(params)

    mlflow.log_metrics({
        "cv_mean_f1_macro": cv_scores["test_f1_macro"].mean(),
        "cv_std_f1_macro": cv_scores["test_f1_macro"].std(),
        "cv_mean_accuracy": cv_scores["test_accuracy"].mean(),
        "cv_mean_balanced_accuracy": cv_scores["test_balanced_accuracy"].mean(),
        "cv_mean_recall_macro": cv_scores["test_recall_macro"].mean(),
    })

Starting running time...
Starting LGBMClassifier(class_weight='balanced', learning_rate=0.05, max_depth=12,
               min_child_samples=200, n_estimators=600, n_jobs=-1,
               num_leaves=300, random_state=42) cross-validation...
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.040791 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 12437
[LightGBM] [Info] Number of data points in the train set: 716260, number of used features: 60
[LightGBM] [Info] Start training from score -2.564949
[LightGBM] [Info] Start training from score -2.564949
[LightGBM] [Info] Start training from score -2.564949
[LightGBM] [Info] Start training from score -2.564949
[LightGBM] [Info] Start training from score -2.564949
[LightGBM] [Info] Start training from score -2.564949
[LightGBM] [Info] Start training from score -2.564949
[LightGBM] [Info] Start 

### Experiment two: Current Dataset

In [15]:
df_copy_with_class_3_and_4 = preprocessed_df.copy()
df_copy_with_class_3_and_4  = df_copy_with_class_3_and_4.set_index("timestamp")

In [16]:
# Split data into X and Y.
X_with_class_3_4 = df_copy_with_class_3_and_4.drop(columns=["class","Instance_id"])

y_with_class_3_4 = df_copy_with_class_3_and_4["class"]
group_with_class_3_4 = df_copy_with_class_3_and_4["Instance_id"]

In [17]:
params = {
    "n_estimators": 600,
    "learning_rate": 0.05,
    "num_leaves": 300,
    "max_depth": 12,
    "min_child_samples": 200,
    "class_weight": "balanced",
    "n_jobs": -1,
    "verbose" : 0,
    "random_state": 42,
}


model = lgb.LGBMClassifier(**params)
model_name = "LightGBM"

splits = 3

scoring = {
    "f1_macro": "f1_macro",
    "accuracy": "accuracy",
    "recall_macro": "recall_macro",
    "balanced_accuracy": "balanced_accuracy"
}

description = "with_class_3_and_4_dataset"

# -------- CREATE FOLDER TO STORE FEATURE IMPORTANCES ------ #
feature_dir = pathlib.Path(f"Feature_importance_{model_name}")
feature_dir.mkdir(parents=True, exist_ok=True)
with mlflow.start_run(run_name="Lightgbm_tuned_v1_with_class_3_and_4_dataset", description="Trained with class 3 and 4"):
        cv_scores, estimator, y_preds, y = run_model(model,
                                            model_name,
                                            splits,
                                            X_with_class_3_4,
                                            y_with_class_3_4,
                                            group_with_class_3_4,
                                            scoring
                                            )
      #------- Create Feature Importance for each fold -------#
        for i in range(splits):
            model = estimator[i].named_steps[f"{model_name}"]
            feature_path =feature_importance(model,X_with_class_3_4,model_name, description,i, feature_dir)
    
            importance_df = pd.DataFrame({
                "Feature": X_with_class_3_4.columns,
                "Importance": model.feature_importances_
            }).sort_values("Importance", ascending=False)
    
            csv_path = feature_dir / f"lightgbm_{description}_feature_importances_v{i+1}.csv"
    
            # Save ALL 72 features
            importance_df.to_csv(
                csv_path,
                index=False
            )
    
        #------- Save Feature Importance plot in Mlflow run ------#
        mlflow.log_artifact(str(feature_path), artifact_path="feature_importance")
            
        # ---------  Confusion Matrix plot ------ #
        cm_path = ConfusionMatrix(y, y_preds, model_name, description)
        mlflow.log_artifact(cm_path)
            
        mlflow.log_params(params)
    
        mlflow.log_metrics({
            "cv_mean_f1_macro": cv_scores["test_f1_macro"].mean(),
            "cv_std_f1_macro": cv_scores["test_f1_macro"].std(),
            "cv_mean_accuracy": cv_scores["test_accuracy"].mean(),
            "cv_mean_balanced_accuracy": cv_scores["test_balanced_accuracy"].mean(),
            "cv_mean_recall_macro": cv_scores["test_recall_macro"].mean(),
        })

Starting running time...
Starting LGBMClassifier(class_weight='balanced', learning_rate=0.05, max_depth=12,
               min_child_samples=200, n_estimators=600, n_jobs=-1,
               num_leaves=300, random_state=42, verbose=0) cross-validation...
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -

### Experiment three: Dataset without permanent classes

In [10]:
df_copy_with_transient_classes = preprocessed_df.copy()
df_copy_with_transient_classes = df_copy_with_transient_classes.set_index("timestamp")

df_copy_with_transient_classes.sort_index()

classes = [1,2,3,4,5,6,7,8]
df_copy_with_transient_classes = df_copy_with_transient_classes[~df_copy_with_transient_classes["class"].isin(classes)]



In [37]:
X_with_transient = df_copy_with_transient_classes.drop(columns=["class","Instance_id"])
y_with_transient = df_copy_with_transient_classes["class"]
group_with_transient = df_copy_with_transient_classes["Instance_id"]

In [46]:
params = {
    "n_estimators": 600,
    "learning_rate": 0.05,
    "num_leaves": 300,
    "max_depth": 12,
    "min_child_samples": 200,
    "class_weight": "balanced",
    "n_jobs": -1,
    "verbose" : 0,
    "random_state": 42,
}


model = lgb.LGBMClassifier(**params)
model_name = "LightGBM"

splits = 3

scoring = {
    "f1_macro": "f1_macro",
    "accuracy": "accuracy",
    "recall_macro": "recall_macro",
    "balanced_accuracy": "balanced_accuracy"
}

description = "transient_class_only"

# -------- CREATE FOLDER TO STORE FEATURE IMPORTANCES ------ #
feature_dir = pathlib.Path(f"Feature_importance_{model_name}")
feature_dir.mkdir(parents=True, exist_ok=True)

with mlflow.start_run(run_name="Lightgbm_tuned_v1_with_transient", description="Trained with transient only"):
        cv_scores, estimator, y_preds, y = run_model(model,
                                            model_name,
                                            splits,
                                            X_with_transient,
                                            y_with_transient,
                                            group_with_transient,
                                            scoring
                                            )
        #------- Create Feature Importance for each fold -------#
        for i in range(splits):
            model = estimator[i].named_steps[f"{model_name}"]
            feature_path =feature_importance(model, X_with_transient,model_name, description,i, feature_dir)
    
            importance_df = pd.DataFrame({
                "Feature": X_with_transient.columns,
                "Importance": model.feature_importances_
            }).sort_values("Importance", ascending=False)
    
            csv_path = feature_dir / f"lightgbm_{description}_feature_importances_v{i+1}.csv"
    
            # Save ALL 72 features
            importance_df.to_csv(
                csv_path,
                index=False
            )
    
        #------- Save Feature Importance plot in Mlflow run ------#
        mlflow.log_artifact(str(feature_path), artifact_path="feature_importance")
            
        # ---------  Confusion Matrix plot ------ #
        cm_path = ConfusionMatrix(y, y_preds, model_name, description)
        mlflow.log_artifact(cm_path)

        # --------- Classfication Report ---------#
        report_dict = classification_report(y, y_preds, target_names=y_with_transient.unique().astype(str), output_dict=True)
        report_lightgbm = pd.DataFrame(report_dict).transpose()

        clf_report_path = f"classification_report_{model_name}_{description}.csv"

        report_lightgbm.to_csv(clf_report_path, index=False)

        mlflow.log_artifact(clf_report_path, artifact_path="classification_report")
            
        mlflow.log_params(params)
    
        mlflow.log_metrics({
            "cv_mean_f1_macro": cv_scores["test_f1_macro"].mean(),
            "cv_std_f1_macro": cv_scores["test_f1_macro"].std(),
            "cv_mean_accuracy": cv_scores["test_accuracy"].mean(),
            "cv_mean_balanced_accuracy": cv_scores["test_balanced_accuracy"].mean(),
            "cv_mean_recall_macro": cv_scores["test_recall_macro"].mean(),
        })

Starting running time...
Starting LGBMClassifier(class_weight='balanced', learning_rate=0.05, max_depth=12,
               min_child_samples=200, n_estimators=600, n_jobs=-1,
               num_leaves=300, random_state=42, verbose=0) cross-validation...
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -

### Observation:

Removing classes 3/4 slightly improves Macro F1 and noticeably improves fold stability, but those classes were not intrinsically difficult. Restricting the problem toward transient detection produces the strongest mean results, although performance becomes much more fold-dependent. Across all formulations, LightGBM consistently relies on temporal dynamics—especially stability duration, pressure changes, and rolling variability across multiple time scales. The dominant classification difficulty appears to be distinguishing permanent fault states from their related transient stages rather than distinguishing completely unrelated fault families.

## Model 2 : RandomForestClassifier

### Experiment one: Current Dataset

In [20]:
X_rf_with_class_3_4 = df_copy_with_class_3_and_4.drop(columns=["class", "Instance_id"])
y_rf_with_class_3_4 = df_copy_with_class_3_and_4["class"]

group_rf_with_class_3_4 = df_copy_with_class_3_and_4["Instance_id"]

In [ ]:
model = RandomForestClassifier(n_jobs=-1, n_estimators=600, min_samples_split=10, class_weight="balanced")

params = {
    "n_estimators": 600,
    "min_samples_split": 10,
    "class_weight": "balanced",
    "n_jobs": -1,
}

model_name = "RandomForest"

splits = 3
scoring = {
    "f1_macro": "f1_macro",
    "accuracy": "accuracy",
    "recall_macro": "recall_macro",
    "balanced_accuracy": "balanced_accuracy"
}

description = "rf_with_dataset_3_4"

# -------- CREATE FOLDER TO STORE FEATURE IMPORTANCES ------ #
feature_dir = pathlib.Path(f"Feature_importance_{model_name}")
feature_dir.mkdir(parents=True, exist_ok=True)

with mlflow.start_run(run_name="Randomforest_v1_with_dataset_3_and_4", description="Random Forest on dataset"):
        cv_scores, estimator, y_preds, y = run_model(model,
                                            model_name,
                                            splits,
                                            X_rf_with_class_3_4, 
                                            y_rf_with_class_3_4, 
                                            group_rf_with_class_3_4, 
                                            scoring
                                            )
        #------- Create Feature Importance for each fold -------#
        for i in range(splits):
            model = estimator[i].named_steps[f"{model_name}"]
            feature_path =feature_importance(model,X_rf_with_class_3_4,model_name, description,i, feature_dir)
    
            importance_df = pd.DataFrame({
                "Feature": X_rf_with_class_3_4.columns,
                "Importance": model.feature_importances_
            }).sort_values("Importance", ascending=False)
    
            csv_path = feature_dir / f"randomforest_{description}_feature_importances_v{i+1}.csv"
    
            # Save ALL 72 features
            importance_df.to_csv(
                csv_path,
                index=False
            )
    
        #------- Save Feature Importance plot in Mlflow run ------#
        mlflow.log_artifact(str(feature_path), artifact_path="feature_importance")
            
        # ---------  Confusion Matrix plot ------ #
        cm_path = ConfusionMatrix(y, y_preds, model_name, description)
        mlflow.log_artifact(cm_path)
        
        
        mlflow.log_params(params)
    
        mlflow.log_metrics({
            "cv_mean_f1_macro": cv_scores["test_f1_macro"].mean(),
            "cv_std_f1_macro": cv_scores["test_f1_macro"].std(),
            "cv_mean_accuracy": cv_scores["test_accuracy"].mean(),
            "cv_mean_balanced_accuracy": cv_scores["test_balanced_accuracy"].mean(),
            "cv_mean_recall_macro": cv_scores["test_recall_macro"].mean(),
        })


Starting running time...
Starting RandomForestClassifier(class_weight='balanced', min_samples_split=10,
                       n_estimators=600, n_jobs=-1) cross-validation...
Macro F1 Scores (Test): [0.77654455 0.76386583 0.81793289]
Accuracy score of Test set: [0.9100463  0.948707   0.93643558]
Accuracy score of Training set: [0.99967973 0.99971304 0.99974069]
Balanced Accuracy score (Test): [0.7641679  0.75481654 0.81293617]
Recall Macro score (Test): [0.7641679  0.75481654 0.81293617]
Average Test Score (Test): 78.61%
Execution time: 1195.5896481670206 seconds


### Experiment two: Dataset without class 3 and 4

In [22]:
X_rf_df_copy_without_class_3_4 = df_copy_without_class_3_4.drop(columns=["class", "Instance_id"])
y_rf_df_copy_without_class_3_4 = df_copy_without_class_3_4["class"]

group_rf_df_copy_without_class_3_4 = df_copy_without_class_3_4["Instance_id"]

In [23]:
model = RandomForestClassifier(n_jobs=-1, n_estimators=600, min_samples_split=10, class_weight="balanced")

params = {
    "n_estimators": 600,
    "min_samples_split": 10,
    "class_weight": "balanced",
    "n_jobs": -1,
}

model_name = "RandomForest"

splits = 3
scoring = {
    "f1_macro": "f1_macro",
    "accuracy": "accuracy",
    "recall_macro": "recall_macro",
    "balanced_accuracy": "balanced_accuracy"
}

description = "rf_without_dataset_3_4"

# -------- CREATE FOLDER TO STORE FEATURE IMPORTANCES ------ #
feature_dir = pathlib.Path(f"Feature_importance_{model_name}")
feature_dir.mkdir(parents=True, exist_ok=True)

with mlflow.start_run(run_name="Randomforest_v1_without_dataset_3_and_4", description="Random Forest on dataset"):
        cv_scores, estimator, y_preds, y = run_model(model,
                                            model_name,
                                            splits,
                                            X_rf_df_copy_without_class_3_4, 
                                            y_rf_df_copy_without_class_3_4, 
                                            group_rf_df_copy_without_class_3_4, 
                                            scoring
                                            )
        #------- Create Feature Importance for each fold -------#
        for i in range(splits):
            model = estimator[i].named_steps[f"{model_name}"]
            feature_path =feature_importance(model, X_rf_df_copy_without_class_3_4,model_name, description,i, feature_dir)
    
            importance_df = pd.DataFrame({
                "Feature": X_rf_df_copy_without_class_3_4.columns,
                "Importance": model.feature_importances_
            }).sort_values("Importance", ascending=False)
    
            csv_path = feature_dir / f"randomforest_{description}_feature_importances_v{i+1}.csv"
    
            # Save ALL 72 features
            importance_df.to_csv(
                csv_path,
                index=False
            )
    
        #------- Save Feature Importance plot in Mlflow run ------#
        mlflow.log_artifact(str(feature_path), artifact_path="feature_importance")
            
        # ---------  Confusion Matrix plot ------ #
        cm_path = ConfusionMatrix(y, y_preds, model_name, description)
        mlflow.log_artifact(cm_path)
            
        mlflow.log_params(params)
    
        mlflow.log_metrics({
            "cv_mean_f1_macro": cv_scores["test_f1_macro"].mean(),
            "cv_std_f1_macro": cv_scores["test_f1_macro"].std(),
            "cv_mean_accuracy": cv_scores["test_accuracy"].mean(),
            "cv_mean_balanced_accuracy": cv_scores["test_balanced_accuracy"].mean(),
            "cv_mean_recall_macro": cv_scores["test_recall_macro"].mean(),
        })


Starting running time...
Starting RandomForestClassifier(class_weight='balanced', min_samples_split=10,
                       n_estimators=600, n_jobs=-1) cross-validation...
Macro F1 Scores (Test): [0.74726584 0.82556595 0.80636949]
Accuracy score of Test set: [0.89863373 0.95954947 0.95014647]
Accuracy score of Training set: [0.9996007  0.99967239 0.99968555]
Balanced Accuracy score (Test): [0.73429012 0.82601118 0.8035715 ]
Recall Macro score (Test): [0.73429012 0.82601118 0.8035715 ]
Average Test Score (Test): 79.31%
Execution time: 854.74375891604 seconds


### Experiment three: Dataset without permanent classes

In [18]:
X_rf_transient = df_copy_with_transient_classes.drop(columns=["Instance_id", "class"])
y_rf_transient = df_copy_with_transient_classes["class"]

group_rf_transient = df_copy_with_transient_classes["Instance_id"]

In [21]:
model = RandomForestClassifier(n_jobs=-1, n_estimators=600, min_samples_split=10, class_weight="balanced")

params = {
    "n_estimators": 600,
    "min_samples_split": 10,
    "class_weight": custom_weight_transient,
    "n_jobs": -1,
}

model_name = "RandomForest"

splits = 3
scoring = {
    "f1_macro": "f1_macro",
    "accuracy": "accuracy",
    "recall_macro": "recall_macro",
    "balanced_accuracy": "balanced_accuracy"
}

description = "rf_with_transient"

# -------- CREATE FOLDER TO STORE FEATURE IMPORTANCES ------ #
feature_dir = pathlib.Path(f"Feature_importance_{model_name}")
feature_dir.mkdir(parents=True, exist_ok=True)

with mlflow.start_run(run_name="Randomforest_v1_with_transient", description="Random Forest on dataset"):
        cv_scores, estimator, y_preds, y = run_model(model,
                                            model_name,
                                            splits,
                                            X_rf_transient, 
                                            y_rf_transient, 
                                            group_rf_transient, 
                                            scoring
                                            )
        #------- Create Feature Importance for each fold -------#
        for i in range(splits):
            model = estimator[i].named_steps[f"{model_name}"]
            feature_path =feature_importance(model,X_rf_transient,model_name, description,i, feature_dir)
    
            importance_df = pd.DataFrame({
                "Feature": X_rf_transient.columns,
                "Importance": model.feature_importances_
            }).sort_values("Importance", ascending=False)
    
            csv_path = feature_dir / f"randomforest_{description}_feature_importances_v{i+1}.csv"
    
            # Save ALL 72 features
            importance_df.to_csv(
                csv_path,
                index=False
            )
    
        #------- Save Feature Importance plot in Mlflow run ------#
        mlflow.log_artifact(str(feature_path), artifact_path="feature_importance")
            
        # ---------  Confusion Matrix plot ------ #
        cm_path = ConfusionMatrix(y, y_preds, model_name, description)
        mlflow.log_artifact(cm_path)

        # --------- Classfication Report ---------#
        report_dict = classification_report(y, y_preds, target_names=y_with_transient.unique().astype(str), output_dict=True)
        report_rf = pd.DataFrame(report_dict).transpose()

        clf_report_path = f"classification_report_{model_name}_{description}.csv"
        
        report_lightgbm.to_csv(clf_report_path, index=False)
            
        mlflow.log_params(params)
    
        mlflow.log_metrics({
            "cv_mean_f1_macro": cv_scores["test_f1_macro"].mean(),
            "cv_std_f1_macro": cv_scores["test_f1_macro"].std(),
            "cv_mean_accuracy": cv_scores["test_accuracy"].mean(),
            "cv_mean_balanced_accuracy": cv_scores["test_balanced_accuracy"].mean(),
            "cv_mean_recall_macro": cv_scores["test_recall_macro"].mean(),
        })


Starting running time...
Starting RandomForestClassifier(class_weight='balanced', min_samples_split=10,
                       n_estimators=600, n_jobs=-1) cross-validation...
Macro F1 Scores (Test): [0.83205004 0.83583876 0.83864753]
Accuracy score of Test set: [0.90828432 0.97082735 0.9772477 ]
Accuracy score of Training set: [0.99995831 0.99995829 0.99995574]
Balanced Accuracy score (Test): [0.8100873  0.8365189  0.83352594]
Recall Macro score (Test): [0.8100873  0.8365189  0.83352594]
Average Test Score (Test): 83.55%
Execution time: 362.48843341699103 seconds


### observation:
Random Forest reproduces most of the same class-level difficulties seen with LightGBM, particularly classes 7, 106, and 107, which suggests those confusions are rooted in the dataset/problem rather than one algorithm. It performs consistently below LightGBM in Macro F1 and shows an extremely large train/test fit gap. Its feature-importance profile is more distributed than LightGBM's and relies more heavily on raw/rolling temperature levels and, notably, calendar-month features. The transient-focused formulation again produces the strongest overall results, but class-specific weaknesses remain hidden beneath the high aggregate accuracy.

## MODEL 3 : XGBOOST

In [25]:
custom_weight_xgb = {
    0:   1.0,
    1: 1.0,
    2: 2.0,
    3: 1.0,
    4: 4.0,
    5: 1.0,
    6: 1.0,
}

### Experiment one: Current Dataset

In [65]:
X_xgb_with_class_3_4 = df_copy_with_class_3_and_4.drop(columns=["class", "Instance_id"])
y_xgb_with_class_3_4 = df_copy_with_class_3_and_4["class"]

group_xgb_with_class_3_4 = df_copy_with_class_3_and_4["Instance_id"]

In [66]:
#--------- Encode Class Labels ---------#
le_xgb_with_class_3_4 = LabelEncoder()

y_xgb_with_class_3_4_encoded = pd.Series(
    le_xgb_with_class_3_4.fit_transform(
        y_xgb_with_class_3_4
    ),
    index=y_xgb_with_class_3_4.index,
    name="class"
)

print(
    dict(
        zip(
            le_xgb_with_class_3_4.classes_,
            le_xgb_with_class_3_4.transform(
                le_xgb_with_class_3_4.classes_
            )
        )
    )
)



{np.float64(0.0): np.int64(0), np.float64(1.0): np.int64(1), np.float64(2.0): np.int64(2), np.float64(3.0): np.int64(3), np.float64(4.0): np.int64(4), np.float64(5.0): np.int64(5), np.float64(6.0): np.int64(6), np.float64(7.0): np.int64(7), np.float64(8.0): np.int64(8), np.float64(101.0): np.int64(9), np.float64(102.0): np.int64(10), np.float64(105.0): np.int64(11), np.float64(106.0): np.int64(12), np.float64(107.0): np.int64(13), np.float64(108.0): np.int64(14)}


In [67]:
#--------- Compute Sample Weights ---------#
weights_xgb_with_class_3_4 = (
    compute_sample_weight(
        class_weight="balanced",
        y=y_xgb_with_class_3_4_encoded
    )
)

In [68]:
params_xgb_with_class_3_4 = {
    "objective": "multi:softprob",
    "tree_method": "hist",
    "num_class": len(
        le_xgb_with_class_3_4.classes_
    ),
    "eval_metric": "mlogloss",
    "random_state": 42
}


XGB_with_class_3_4 = xgb.XGBClassifier(
    **params_xgb_with_class_3_4
)

model_name = "XGBoost"

description = "with_class_3_and_4_dataset"

In [ ]:
# -------- CREATE FOLDER TO STORE FEATURE IMPORTANCES ------ #
feature_dir = pathlib.Path(f"Feature_importance_XGBoost")
feature_dir.mkdir(parents=True, exist_ok=True)

splits = 3

with mlflow.start_run(
    run_name="XGBoost_with_class_3_and_4",
    description="XGBoost trained with classes 3 and 4"
):

    cv_scores_xgb_1, estimators_xgb_1, y_preds_xgb_1, y_true_xgb_1 = (
        run_model(
            XGB_with_class_3_4,
            model_name,
            splits,
            X_xgb_with_class_3_4,
            y_xgb_with_class_3_4_encoded,
            group_xgb_with_class_3_4,
            scoring,
            weights_xgb_with_class_3_4
        )
    )

    #------- Create Feature Importance for each fold -------#
    for i in range(splits):
        model = estimators_xgb_1[i].named_steps[f"{model_name}"]
        feature_path =feature_importance(model,X_xgb_with_class_3_4,model_name, description,i, feature_dir)

        importance_df = pd.DataFrame({
            "Feature": X_xgb_with_class_3_4.columns,
            "Importance": model.feature_importances_
        }).sort_values("Importance", ascending=False)

        csv_path = feature_dir / f"xgboost_{description}_feature_importances_v{i+1}.csv"

        # Save ALL 72 features
        importance_df.to_csv(
            csv_path,
            index=False
        )

    #------- Save Feature Importance plot in Mlflow run ------#
    mlflow.log_artifact(str(feature_path), artifact_path="feature_importance")


    y_true_xgb_1_original = (
    le_xgb_with_class_3_4.inverse_transform(
        np.asarray(y_true_xgb_1).astype(int)
    )
    )

    y_preds_xgb_1_original = (
        le_xgb_with_class_3_4.inverse_transform(
            np.asarray(y_preds_xgb_1).astype(int)
        )
    )

    cmf = ConfusionMatrix(y_true_xgb_1_original, y_preds_xgb_1_original, "XGBoost", "with_class_3_4")
    mlflow.log_artifact(cm_path)

    mlflow.log_params(
        params_xgb_with_class_3_4
    )

    mlflow.log_metrics({
        "cv_mean_f1_macro":
            cv_scores_xgb_1["test_f1_macro"].mean(),

        "cv_std_f1_macro":
            cv_scores_xgb_1["test_f1_macro"].std(),

        "cv_mean_accuracy":
            cv_scores_xgb_1["test_accuracy"].mean(),

        "cv_mean_balanced_accuracy":
            cv_scores_xgb_1[
                "test_balanced_accuracy"
            ].mean(),

        "cv_mean_recall_macro":
            cv_scores_xgb_1[
                "test_recall_macro"
            ].mean()
    })

Starting running time...
Starting XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=True, eval_metric='mlogloss',
              feature_types=None, feature_weights=None, gamma=None,
              grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None, num_class=15, ...) cross-validation...
Macro F1 Scores (Test): [0.73851588 0.23682496 0.72966529]
Accuracy score of Test set: [0.95309174 0.53977023 0.93983887]
Accuracy score of Training set: [0.99833577 0.53691117 0.96441266]
B

### Experiment two: Without classes 3 and 4

In [70]:
X_xgb_without_class_3_4 = df_copy_without_class_3_4.drop(columns=["class","Instance_id"])

y_xgb_without_class_3_4 = df_copy_without_class_3_4["class"]

group_xgb_without_class_3_4 = df_copy_without_class_3_4["Instance_id"]


In [71]:
#--------- Encode Labels ---------#
le_xgb_without_class_3_4 = LabelEncoder()

y_xgb_without_class_3_4_encoded = pd.Series(
    le_xgb_without_class_3_4.fit_transform(
        y_xgb_without_class_3_4
    ),
    index=y_xgb_without_class_3_4.index,
    name="class"
)


In [72]:
#--------- Compute Sample Weights ---------#
weights_xgb_without_class_3_4 = (
    compute_sample_weight(
        class_weight="balanced",
        y=y_xgb_without_class_3_4_encoded
    )
)

In [73]:
params_xgb_without_class_3_4 = {
    "objective": "multi:softprob",
    "tree_method": "hist",
    "num_class": len(
        le_xgb_without_class_3_4.classes_
    ),
    "eval_metric": "mlogloss",
    "random_state": 42
}


XGB_without_class_3_4 = xgb.XGBClassifier(
    **params_xgb_without_class_3_4
)
model_name = "XGBoost"
description = "without_class_3_4"

In [ ]:
# -------- CREATE FOLDER TO STORE FEATURE IMPORTANCES ------ #
feature_dir = pathlib.Path(f"Feature_importance_XGBoost")
feature_dir.mkdir(parents=True, exist_ok=True)

splits = 3

with mlflow.start_run(run_name="XGBoost_without_class_3_and_4", description="XGBoost trained without classes 3 and 4"):
    cv_scores_xgb_2, estimators_xgb_2, y_preds_xgb_2, y_true_xgb_2 = (
        run_model(
            XGB_without_class_3_4,
            "XGBoost",
            splits,
            X_xgb_without_class_3_4,
            y_xgb_without_class_3_4_encoded,
            group_xgb_without_class_3_4,
            scoring,
            weights_xgb_without_class_3_4
        )
    )
    #------- Create Feature Importance for each fold -------#
    for i in range(splits):
        model = estimators_xgb_2[i].named_steps[f"{model_name}"]
        feature_path =feature_importance(model,X_xgb_without_class_3_4,model_name, description,i, feature_dir)

        importance_df = pd.DataFrame({
            "Feature": X_xgb_without_class_3_4.columns,
            "Importance": model.feature_importances_
        }).sort_values("Importance", ascending=False)

        csv_path = feature_dir / f"xgboost_{description}_feature_importances_v{i+1}.csv"

        # Save ALL 72 features
        importance_df.to_csv(
            csv_path,
            index=False
        )

    #------- Save Feature Importance plot in Mlflow run ------#
    mlflow.log_artifact(str(feature_path), artifact_path="feature_importance")


    #------- Transform the encoded labels back to original ------#
    y_true_xgb_2_original = (
    le_xgb_without_class_3_4.inverse_transform(
        np.asarray(y_true_xgb_2).astype(int)
    )
    )

    y_preds_xgb_2_original = (
        le_xgb_without_class_3_4.inverse_transform(
            np.asarray(y_preds_xgb_2).astype(int)
        )
    )

    #------- Plot Confusion Matrix ------#
    cmf = ConfusionMatrix(y_true_xgb_2_original, y_preds_xgb_2_original, "XGBoost", "without_class_3_4")
    mlflow.log_artifact(cm_path)

    mlflow.log_params(
        params_xgb_without_class_3_4
    )

    mlflow.log_metrics({
        "cv_mean_f1_macro": cv_scores_xgb_2["test_f1_macro"].mean(),
        "cv_std_f1_macro":cv_scores_xgb_2["test_f1_macro"].std(),
        "cv_mean_accuracy":cv_scores_xgb_2["test_accuracy"].mean(),
        "cv_mean_balanced_accuracy":cv_scores_xgb_2["test_balanced_accuracy"].mean(),
        "cv_mean_recall_macro": cv_scores_xgb_2["test_recall_macro"].mean()
    })

Starting running time...
Starting XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=True, eval_metric='mlogloss',
              feature_types=None, feature_weights=None, gamma=None,
              grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None, num_class=13, ...) cross-validation...
Macro F1 Scores (Test): [0.81604581 0.60959139 0.60783861]
Accuracy score of Test set: [0.94882891 0.90130037 0.91007357]
Accuracy score of Training set: [0.99995114 0.93465224 0.93143789]
B

### Experiment 3 — Transient fault classes only

In [48]:
X_xgb_transient = df_copy_with_transient_classes.drop(columns=["class","Instance_id"])

y_xgb_transient = df_copy_with_transient_classes["class"]

group_xgb_transient = df_copy_with_transient_classes["Instance_id"]


In [49]:
#--------- Encode Labels ---------#
le_xgb_transient = LabelEncoder()

y_xgb_transient_encoded = pd.Series(
    le_xgb_transient.fit_transform(
        y_xgb_transient
    ),
    index=y_xgb_transient.index,
    name="class"
)


print(
    dict(
        zip(
            le_xgb_transient.classes_,
            le_xgb_transient.transform(
                le_xgb_transient.classes_
            )
        )
    )
)


{np.float64(0.0): np.int64(0), np.float64(101.0): np.int64(1), np.float64(102.0): np.int64(2), np.float64(105.0): np.int64(3), np.float64(106.0): np.int64(4), np.float64(107.0): np.int64(5), np.float64(108.0): np.int64(6)}


In [50]:
#--------- Compute Sample Weights ---------#
weights_xgb_transient = (
    compute_sample_weight(
        class_weight="balanced",
        y=y_xgb_transient_encoded
    )
)


params_xgb_transient = {
    "objective": "multi:softprob",
    "tree_method": "hist",
    "num_class": len(
        le_xgb_transient.classes_
    ),
    "eval_metric": "mlogloss",
    "random_state": 42
}


XGB_transient = xgb.XGBClassifier(**params_xgb_transient)

model_name = "XGBoost"

description = "transient_class"

In [51]:
# -------- CREATE FOLDER TO STORE FEATURE IMPORTANCES ------ #
feature_dir = pathlib.Path(f"Feature_importance_XGBoost")
feature_dir.mkdir(parents=True, exist_ok=True)

splits = 3

with mlflow.start_run(run_name="XGBoost_transient_only", description=("XGBoost trained on normal operation and transient fault classes only")):
    cv_scores_xgb_3, estimators_xgb_3, y_preds_xgb_3, y_true_xgb_3 = (
        run_model(
            XGB_transient,
            "XGBoost",
            splits,
            X_xgb_transient,
            y_xgb_transient_encoded,
            group_xgb_transient,
            scoring,
            weights_xgb_transient
        )
    )

    #------- Create Feature Importance for each fold -------#
    for i in range(splits):
        model = estimators_xgb_3[i].named_steps[f"{model_name}"]
        feature_path =feature_importance(model,X_xgb_transient,model_name, description,i, feature_dir)

        importance_df = pd.DataFrame({
            "Feature": X_xgb_transient.columns,
            "Importance": model.feature_importances_
        }).sort_values("Importance", ascending=False)

        csv_path = feature_dir / f"xgboost_{description}_feature_importances_v{i+1}.csv"

        # Save ALL 72 features
        importance_df.to_csv(
            csv_path,
            index=False
        )

    #------- Save Feature Importance plot in Mlflow run ------#
    mlflow.log_artifact(str(feature_path), artifact_path="feature_importance")


    #------- Transform the encoded labels back to original ------#
    y_true_xgb_3_original = (
    le_xgb_transient.inverse_transform(
        np.asarray(y_true_xgb_3).astype(int)
    )
    )

    y_preds_xgb_3_original = (
        le_xgb_transient.inverse_transform(
            np.asarray(y_preds_xgb_3).astype(int)
        )
    )

    #------- Plot Confusion Matrix ------#
    cmf = ConfusionMatrix(y_true_xgb_3_original, y_preds_xgb_3_original, "XGBoost", "transient")
    mlflow.log_artifact(cm_path)

    # --------- Classfication Report ---------#
    report_dict = classification_report(y_true_xgb_3_original, y_preds_xgb_3_original, target_names=y_xgb_transient.unique().astype(str), output_dict=True)
    report_xgb = pd.DataFrame(report_dict).transpose()

    clf_report_path = f"classification_report_{model_name}_{description}.csv"
    
    report_lightgbm.to_csv(clf_report_path, index=False)

    mlflow.log_params(
        params_xgb_transient
    )

    mlflow.log_metrics({
        "cv_mean_f1_macro": cv_scores_xgb_3["test_f1_macro"].mean(),
        "cv_std_f1_macro": cv_scores_xgb_3["test_f1_macro"].std(),
        "cv_mean_accuracy": cv_scores_xgb_3["test_accuracy"].mean(),
        "cv_mean_balanced_accuracy": cv_scores_xgb_3["test_balanced_accuracy"].mean(),
        "cv_mean_recall_macro": cv_scores_xgb_3["test_recall_macro"].mean()
    })

Starting running time...
Starting XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=True, eval_metric='mlogloss',
              feature_types=None, feature_weights=None, gamma=None,
              grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None, num_class=7, ...) cross-validation...
Macro F1 Scores (Test): [0.89517641 0.77377954 0.82595813]
Accuracy score of Test set: [0.97564877 0.93350405 0.9788341 ]
Accuracy score of Training set: [0.99999218 0.99110723 0.99998959]
Ba

### Conclusion:

| Model         |    Current | Without 3/4 | Transient-focused |
| ------------- | ---------: | ----------: | ----------------: |
| LightGBM      | **81.62%** |  **83.37%** |        **88.67%** |
| Random Forest | **78.61%** |  **79.31%** |        **84.59%** |
| XGBoost       | **56.83%** |  **67.78%** |        **85.85%** |


#### LightGBM:

→ strongest overall
→ relatively stable
→ very consistent feature families

#### Random Forest
→ slightly weaker
→ heavy train/test fit gap
→ more reliance on raw/rolling temperature and month

#### XGBoost
→ highly unstable on the broader datasets
→ much stronger once the task is narrowed
→ feature importance changes dramatically between folds

In [30]:
report_lightgbm

,precision,recall,f1-score,support
0.0,0.973320,0.741515,0.841750,242892.000000
101.0,0.907915,0.851331,0.878713,175457.000000
102.0,0.485273,0.617177,0.543334,2189.000000
105.0,0.860233,0.662380,0.748452,35300.000000
106.0,0.004145,0.460526,0.008216,152.000000
107.0,0.495945,0.958712,0.653719,76535.000000
108.0,0.975616,0.707970,0.820518,43177.000000
accuracy,0.795943,0.795943,0.795943,0.795943
macro avg,0.671778,0.714230,0.642100,575702.000000
weighted avg,0.881050,0.795943,0.819350,575702.000000


In [31]:
report_rf

,precision,recall,f1-score,support
0.0,0.924203,0.983470,0.952916,242892.000000
101.0,0.983503,0.985358,0.984430,175457.000000
102.0,0.992853,0.888534,0.937801,2189.000000
105.0,0.993909,0.845864,0.913930,35300.000000
106.0,0.541667,0.427632,0.477941,152.000000
107.0,0.934405,0.807591,0.866382,76535.000000
108.0,1.000000,1.000000,1.000000,43177.000000
accuracy,0.952958,0.952958,0.952958,0.952958
macro avg,0.910077,0.848350,0.876200,575702.000000
weighted avg,0.953751,0.952958,0.951974,575702.000000


In [32]:
report_xgb

,precision,recall,f1-score,support
0.0,0.971145,0.982840,0.976958,242892.000000
101.0,0.991194,0.987969,0.989579,175457.000000
102.0,0.984600,0.876199,0.927242,2189.000000
105.0,0.991719,0.990652,0.991185,35300.000000
106.0,0.054487,0.335526,0.093750,152.000000
107.0,0.956253,0.920794,0.938189,76535.000000
108.0,0.999815,0.999328,0.999571,43177.000000
accuracy,0.977294,0.977294,0.977294,0.977294
macro avg,0.849888,0.870473,0.845211,575702.000000
weighted avg,0.978497,0.977294,0.977796,575702.000000
